abi/codecForCompositeTypes_option.go

```diff
@@ -0,0 +1,65 @@
+package abi
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+func (c *codec) encodeNestedOption(writer io.Writer, value OptionValue) error {
+	if value.Value == nil {
+		_, err := writer.Write([]byte{0})
+		return err
+	}
+
+	_, err := writer.Write([]byte{1})
+	if err != nil {
+		return err
+	}
+
+	return c.doEncodeNested(writer, value.Value)
+}
+
+func (c *codec) encodeTopLevelOption(writer io.Writer, value OptionValue) error {
+	if value.Value == nil {
+		return nil
+	}
+
+	_, err := writer.Write([]byte{1})
+	if err != nil {
+		return err
+	}
+
+	return c.doEncodeNested(writer, value.Value)
+}
+
+func (c *codec) decodeNestedOption(reader io.Reader, value *OptionValue) error {
+	bytes, err := readBytesExactly(reader, 1)
+	if err != nil {
+		return err
+	}
+
+	if bytes[0] == 0 {
+		value.Value = nil
+		return nil
+	}
+
+	return c.doDecodeNested(reader, value.Value)
+}
+
+func (c *codec) decodeTopLevelOption(data []byte, value *OptionValue) error {
+	if len(data) == 0 {
+		value.Value = nil
+		return nil
+	}
+
```

Comment:
You can do here `return c.decodeNestedOption(reader, value)`, and in decodeNestedOption do the logic with 1st byte 0 or 1 then doDecodeNested

Reply:
In `decodeTopLevelOption`, only `0x01` is allowed as a _value presence indicator_, in my understanding. When value is missing, data must be empty. On the other hand, for `decodeNestedOption`, the value presence indicator must be either 0 or 1.

How does this sound? I can do the refactoring if applicable.

Reply:
it's all right